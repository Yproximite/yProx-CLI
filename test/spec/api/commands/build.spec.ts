import { readFileSync, existsSync, statSync } from 'fs';
import { createFakeEnv } from '../../fake-env';

const readFile = (filename: string, charset: string | null = 'utf8') => readFileSync(filename, charset);

const readFixture = (filename: string, charset: string | null = 'utf8') => readFileSync(`${__dirname}/../../__fixtures__/${filename}`, charset);

describe('command: build', () => {
  let oldEnv = process.env;
  beforeEach(() => {
    oldEnv = process.env;
    console.info = jest.fn();
  });
  afterEach(() => {
    process.env = oldEnv;
    // @ts-ignore
    console.info.mockRestore();
  });

  it('should build entries', async () => {
    const { api, cleanup, run } = await createFakeEnv(
      {
        'package.json': readFixture('package.json'),
        'yarn.lock': readFixture('yarn.lock'),
        // rollup
        'src/components/button/Button.vue': readFixture('yarn.lock'),
        'src/components/button/index.js': readFixture('src/components/button/index.js'),
        // css
        'src/css/bar.css': readFixture('src/css/bar.css'),
        'src/css/foo.css': readFixture('src/css/foo.css'),
        // js
        'src/js/bar.js': readFixture('src/js/bar.js'),
        'src/js/foo.js': readFixture('src/js/foo.js'),
        // sass
        'src/sass/_form.scss': readFixture('src/sass/_form.scss'),
        'src/sass/style.scss': readFixture('src/sass/style.scss'),
        // files
        'src/lorem.txt': readFixture('src/lorem.txt'),
        'src/udhr.txt': readFixture('src/udhr.txt'),
        // images
        'src/images/guts-white-hair.png': readFixture('src/images/guts-white-hair.png', null),
        'src/images/jax.jpg': readFixture('src/images/jax.jpg', null),
        'src/images/golfer.gif': readFixture('src/images/golfer.gif', null),
        'src/images/uk.svg': readFixture('src/images/uk.svg', null),
      },
      'production',
      true
    );

    await run('yarn'); // install dependencies
    await api.executeCommand('build'); // we could use `yarn build`, but we won't have access to mocked `console.log`

    // should have built files with Rollup handler
    expect(true).toBeTruthy();
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: rollup :: start bundling "button.js"');
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: rollup :: finished bundling "button.js"');

    expect(existsSync(api.resolve('dist/js/button.js'))).toBeTruthy();
    expect(readFile(api.resolve('dist/js/button.js'))).toMatchSnapshot();

    // should have built files with JS handler
    expect(true).toBeTruthy();
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: js :: start bundling "scripts.js"');
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: js :: finished bundling "scripts.js"');

    expect(existsSync(api.resolve('dist/js/scripts.js'))).toBeTruthy();
    expect(readFile(api.resolve('dist/js/scripts.js'))).toMatchSnapshot();

    // should have built files with CSS
    expect(true).toBeTruthy();
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: css :: start bundling "legacy-styles.css"');
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: css :: finished bundling "legacy-styles.css"');

    expect(existsSync(api.resolve('dist/css/legacy-styles.css'))).toBeTruthy();
    expect(readFile(api.resolve('dist/css/legacy-styles.css'))).toMatchSnapshot();

    // should have built files with Sass handler
    expect(true).toBeTruthy();
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: sass :: start bundling "styles.css"');
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: sass :: finished bundling "styles.css"');

    expect(existsSync(api.resolve('dist/css/styles.css'))).toBeTruthy();
    expect(readFile(api.resolve('dist/css/styles.css'))).toMatchSnapshot();

    // should have copied files
    expect(true).toBeTruthy();
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: file :: start copying "files to copy"');
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: file :: done copying "files to copy"');

    expect(readFile(api.resolve('dist/lorem.txt'))).toEqual(readFile(api.resolve('src/lorem.txt')));
    expect(readFile(api.resolve('dist/lorem.txt'))).toContain('Lorem ipsum dolor sit amet.');

    expect(readFile(api.resolve('dist/udhr.txt'))).toEqual(readFile(api.resolve('src/udhr.txt')));
    expect(readFile(api.resolve('dist/udhr.txt'))).toContain('Universal Declaration of Human Rights - English');

    // should have optimized images
    expect(true).toBeTruthy();
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: image :: start optimizing "images to optimize"');
    expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: image :: done optimizing "images to optimize"');

    expect(statSync(api.resolve('src/images/guts-white-hair.png')).size).toBeGreaterThan(1024 * 1024);
    expect(statSync(api.resolve('dist/images/guts-white-hair.png')).size).toBeLessThan(1024 * 1024);

    expect(statSync(api.resolve('src/images/jax.jpg')).size).toBeGreaterThan(230 * 1024);
    expect(statSync(api.resolve('dist/images/jax.jpg')).size).toBeLessThan(230 * 1024);

    expect(statSync(api.resolve('src/images/golfer.gif')).size).toBeGreaterThan(2800 * 1024);
    expect(statSync(api.resolve('dist/images/golfer.gif')).size).toBeLessThan(435 * 1024);

    expect(statSync(api.resolve('src/images/uk.svg')).size).toBeGreaterThan(55 * 1024);
    expect(statSync(api.resolve('dist/images/uk.svg')).size).toBeLessThan(42 * 1024);

    await cleanup();
  }, 50000);
});
