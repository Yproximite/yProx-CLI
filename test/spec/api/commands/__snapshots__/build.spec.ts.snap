// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command: build lint (and fix) before build should lint (and but fix) files built with handler \`css\`, before building them: css/bar.css after lint 1`] = `
".bar {
  color: blue;
}
"
`;

exports[`command: build lint (and fix) before build should lint (and but fix) files built with handler \`css\`, before building them: css/bar.css before lint 1`] = `
".bar {
  color: blue;;;
}
"
`;

exports[`command: build lint (and fix) before build should lint (and but fix) files built with handler \`js\`, before building them: js/bar.js after lint 1`] = `
"console.log('Hello world from \`bar.js\`!');
"
`;

exports[`command: build lint (and fix) before build should lint (and but fix) files built with handler \`js\`, before building them: js/bar.js before lint 1`] = `
"console.log('Hello world from \`bar.js\`!');;;
"
`;

exports[`command: build lint (and fix) before build should lint (and but fix) files built with handler \`rollup\`, before building them: button/index.js after lint 1`] = `
"import Vue from 'vue';
import Button from './Button';

Vue.component('y-button', Button);

console.log(\\"Hello from index.js!\\");
"
`;

exports[`command: build lint (and fix) before build should lint (and but fix) files built with handler \`rollup\`, before building them: button/index.js before lint 1`] = `
"import Vue from 'vue';
import Button from './Button';

Vue.component('y-button', Button);

console.log(\\"Hello from index.js!\\");;;
"
`;

exports[`command: build lint (and fix) before build should lint (and but fix) files built with handler \`sass\`, before building them: sass/style.scss after lint 1`] = `
"@import \\"~bootstrap/scss/bootstrap-reboot\\";

$font-stack: Helvetica, sans-serif;

body {
  font: 100% $font-stack;
  color: red;
}

@import \\"form\\";
"
`;

exports[`command: build lint (and fix) before build should lint (and but fix) files built with handler \`sass\`, before building them: sass/style.scss before lint 1`] = `
"@import \\"~bootstrap/scss/bootstrap-reboot\\";

$font-stack: Helvetica, sans-serif;

body {
  font: 100% $font-stack;
  color: red;;;
}

@import \\"form\\";
"
`;

exports[`command: build should build entries (development mode): dev css 1`] = `
".foo {
  color: red;;;
}

.bar {
  color: blue;;;
}
"
`;

exports[`command: build should build entries (development mode): dev js 1`] = `
"console.log('Hello world from \`foo.js\`!');

var foo = 'My variable';
console.log((\\"My variable value: \\" + foo));

console.log('Hello world from \`bar.js\`!');;;
"
`;

exports[`command: build should build entries (development mode): dev sass 1`] = `
"/*!
 * Bootstrap Reboot v4.2.1 (https://getbootstrap.com/)
 * Copyright 2011-2018 The Bootstrap Authors
 * Copyright 2011-2018 Twitter, Inc.
 * Licensed under MIT (https://github.com/twbs/bootstrap/blob/master/LICENSE)
 * Forked from Normalize.css, licensed MIT (https://github.com/necolas/normalize.css/blob/master/LICENSE.md)
 */
*,
*::before,
*::after {
  box-sizing: border-box; }

html {
  font-family: sans-serif;
  line-height: 1.15;
  -webkit-text-size-adjust: 100%;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0); }

article, aside, figcaption, figure, footer, header, hgroup, main, nav, section {
  display: block; }

body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", Arial, \\"Noto Sans\\", sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\", \\"Noto Color Emoji\\";
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #212529;
  text-align: left;
  background-color: #fff; }

[tabindex=\\"-1\\"]:focus {
  outline: 0 !important; }

hr {
  box-sizing: content-box;
  height: 0;
  overflow: visible; }

h1, h2, h3, h4, h5, h6 {
  margin-top: 0;
  margin-bottom: 0.5rem; }

p {
  margin-top: 0;
  margin-bottom: 1rem; }

abbr[title],
abbr[data-original-title] {
  text-decoration: underline;
  -webkit-text-decoration: underline dotted;
          text-decoration: underline dotted;
  cursor: help;
  border-bottom: 0;
  text-decoration-skip-ink: none; }

address {
  margin-bottom: 1rem;
  font-style: normal;
  line-height: inherit; }

ol,
ul,
dl {
  margin-top: 0;
  margin-bottom: 1rem; }

ol ol,
ul ul,
ol ul,
ul ol {
  margin-bottom: 0; }

dt {
  font-weight: 700; }

dd {
  margin-bottom: .5rem;
  margin-left: 0; }

blockquote {
  margin: 0 0 1rem; }

b,
strong {
  font-weight: bolder; }

small {
  font-size: 80%; }

sub,
sup {
  position: relative;
  font-size: 75%;
  line-height: 0;
  vertical-align: baseline; }

sub {
  bottom: -.25em; }

sup {
  top: -.5em; }

a {
  color: #007bff;
  text-decoration: none;
  background-color: transparent; }
  a:hover {
    color: #0056b3;
    text-decoration: underline; }

a:not([href]):not([tabindex]) {
  color: inherit;
  text-decoration: none; }
  a:not([href]):not([tabindex]):hover, a:not([href]):not([tabindex]):focus {
    color: inherit;
    text-decoration: none; }
  a:not([href]):not([tabindex]):focus {
    outline: 0; }

pre,
code,
kbd,
samp {
  font-family: SFMono-Regular, Menlo, Monaco, Consolas, \\"Liberation Mono\\", \\"Courier New\\", monospace;
  font-size: 1em; }

pre {
  margin-top: 0;
  margin-bottom: 1rem;
  overflow: auto; }

figure {
  margin: 0 0 1rem; }

img {
  vertical-align: middle;
  border-style: none; }

svg {
  overflow: hidden;
  vertical-align: middle; }

table {
  border-collapse: collapse; }

caption {
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  color: #6c757d;
  text-align: left;
  caption-side: bottom; }

th {
  text-align: inherit; }

label {
  display: inline-block;
  margin-bottom: 0.5rem; }

button {
  border-radius: 0; }

button:focus {
  outline: 1px dotted;
  outline: 5px auto -webkit-focus-ring-color; }

input,
button,
select,
optgroup,
textarea {
  margin: 0;
  font-family: inherit;
  font-size: inherit;
  line-height: inherit; }

button,
input {
  overflow: visible; }

button,
select {
  text-transform: none; }

button,
[type=\\"button\\"],
[type=\\"reset\\"],
[type=\\"submit\\"] {
  -webkit-appearance: button; }

button::-moz-focus-inner,
[type=\\"button\\"]::-moz-focus-inner,
[type=\\"reset\\"]::-moz-focus-inner,
[type=\\"submit\\"]::-moz-focus-inner {
  padding: 0;
  border-style: none; }

input[type=\\"radio\\"],
input[type=\\"checkbox\\"] {
  box-sizing: border-box;
  padding: 0; }

input[type=\\"date\\"],
input[type=\\"time\\"],
input[type=\\"datetime-local\\"],
input[type=\\"month\\"] {
  -webkit-appearance: listbox; }

textarea {
  overflow: auto;
  resize: vertical; }

fieldset {
  min-width: 0;
  padding: 0;
  margin: 0;
  border: 0; }

legend {
  display: block;
  width: 100%;
  max-width: 100%;
  padding: 0;
  margin-bottom: .5rem;
  font-size: 1.5rem;
  line-height: inherit;
  color: inherit;
  white-space: normal; }

progress {
  vertical-align: baseline; }

[type=\\"number\\"]::-webkit-inner-spin-button,
[type=\\"number\\"]::-webkit-outer-spin-button {
  height: auto; }

[type=\\"search\\"] {
  outline-offset: -2px;
  -webkit-appearance: none; }

[type=\\"search\\"]::-webkit-search-decoration {
  -webkit-appearance: none; }

::-webkit-file-upload-button {
  font: inherit;
  -webkit-appearance: button; }

output {
  display: inline-block; }

summary {
  display: list-item;
  cursor: pointer; }

template {
  display: none; }

[hidden] {
  display: none !important; }

body {
  font: 100% Helvetica, sans-serif;
  color: red; }

input {
  font-size: 20px; }
"
`;

exports[`command: build should build entries (production mode): prod css 1`] = `
".foo{color:red}.bar{color:#00f}
/*# sourceMappingURL=legacy-styles.css.map */
"
`;

exports[`command: build should build entries (production mode): prod js 1`] = `
"console.log(\\"Hello world from \`foo.js\`!\\");var foo=\\"My variable\\";console.log(\\"My variable value: \\"+foo),console.log(\\"Hello world from \`bar.js\`!\\");
//# sourceMappingURL=scripts.js.map
"
`;

exports[`command: build should build entries (production mode): prod sass 1`] = `
"/*!
 * Bootstrap Reboot v4.2.1 (https://getbootstrap.com/)
 * Copyright 2011-2018 The Bootstrap Authors
 * Copyright 2011-2018 Twitter, Inc.
 * Licensed under MIT (https://github.com/twbs/bootstrap/blob/master/LICENSE)
 * Forked from Normalize.css, licensed MIT (https://github.com/necolas/normalize.css/blob/master/LICENSE.md)
 */*,:after,:before{box-sizing:border-box}html{font-family:sans-serif;line-height:1.15;-webkit-text-size-adjust:100%;-webkit-tap-highlight-color:rgba(0,0,0,0)}article,aside,figcaption,figure,footer,header,hgroup,main,nav,section{display:block}body{margin:0;font-family:-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,Noto Sans,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol,Noto Color Emoji;font-size:1rem;font-weight:400;line-height:1.5;color:#212529;text-align:left;background-color:#fff}[tabindex=\\"-1\\"]:focus{outline:0!important}hr{box-sizing:content-box;height:0;overflow:visible}h1,h2,h3,h4,h5,h6{margin-top:0;margin-bottom:.5rem}p{margin-top:0;margin-bottom:1rem}abbr[data-original-title],abbr[title]{text-decoration:underline;-webkit-text-decoration:underline dotted;text-decoration:underline dotted;cursor:help;border-bottom:0;text-decoration-skip-ink:none}address{font-style:normal;line-height:inherit}address,dl,ol,ul{margin-bottom:1rem}dl,ol,ul{margin-top:0}ol ol,ol ul,ul ol,ul ul{margin-bottom:0}dt{font-weight:700}dd{margin-bottom:.5rem;margin-left:0}blockquote{margin:0 0 1rem}b,strong{font-weight:bolder}small{font-size:80%}sub,sup{position:relative;font-size:75%;line-height:0;vertical-align:baseline}sub{bottom:-.25em}sup{top:-.5em}a{color:#007bff;text-decoration:none;background-color:transparent}a:hover{color:#0056b3;text-decoration:underline}a:not([href]):not([tabindex]),a:not([href]):not([tabindex]):focus,a:not([href]):not([tabindex]):hover{color:inherit;text-decoration:none}a:not([href]):not([tabindex]):focus{outline:0}code,kbd,pre,samp{font-family:SFMono-Regular,Menlo,Monaco,Consolas,Liberation Mono,Courier New,monospace;font-size:1em}pre{margin-top:0;margin-bottom:1rem;overflow:auto}figure{margin:0 0 1rem}img{border-style:none}img,svg{vertical-align:middle}svg{overflow:hidden}table{border-collapse:collapse}caption{padding-top:.75rem;padding-bottom:.75rem;color:#6c757d;text-align:left;caption-side:bottom}th{text-align:inherit}label{display:inline-block;margin-bottom:.5rem}button{border-radius:0}button:focus{outline:1px dotted;outline:5px auto -webkit-focus-ring-color}button,input,optgroup,select,textarea{margin:0;font-family:inherit;font-size:inherit;line-height:inherit}button,input{overflow:visible}button,select{text-transform:none}[type=button],[type=reset],[type=submit],button{-webkit-appearance:button}[type=button]::-moz-focus-inner,[type=reset]::-moz-focus-inner,[type=submit]::-moz-focus-inner,button::-moz-focus-inner{padding:0;border-style:none}input[type=checkbox],input[type=radio]{box-sizing:border-box;padding:0}input[type=date],input[type=datetime-local],input[type=month],input[type=time]{-webkit-appearance:listbox}textarea{overflow:auto;resize:vertical}fieldset{min-width:0;padding:0;margin:0;border:0}legend{display:block;width:100%;max-width:100%;padding:0;margin-bottom:.5rem;font-size:1.5rem;line-height:inherit;color:inherit;white-space:normal}progress{vertical-align:baseline}[type=number]::-webkit-inner-spin-button,[type=number]::-webkit-outer-spin-button{height:auto}[type=search]{outline-offset:-2px;-webkit-appearance:none}[type=search]::-webkit-search-decoration{-webkit-appearance:none}::-webkit-file-upload-button{font:inherit;-webkit-appearance:button}output{display:inline-block}summary{display:list-item;cursor:pointer}template{display:none}[hidden]{display:none!important}body{font:100% Helvetica,sans-serif;color:red}input{font-size:20px}
/*# sourceMappingURL=styles.css.map */
"
`;
